ca/server.go

```diff
@@ -808,3 +826,162 @@ func isFinalState(status api.IssuanceStatus) bool {
 
 	return false
 }
+
+func (s *Server) finishRootRotation(expectedRootCA *api.RootCA) error {
+	// when the nodes are all converged on having TLS certificates issued by the correct entity, we can
+	// try to finish the root rotation
+	clusterID := s.securityConfig.ClientTLSCreds.Organization()
+	return s.store.Update(func(tx store.Tx) error {
+		cluster := store.GetCluster(tx, clusterID)
+		if cluster == nil {
+			return fmt.Errorf("unable to get cluster %s", clusterID)
+		}
+
+		// If the RootCA object has changed (because another root rotation was started or because some other node
+		// had finished the root rotation), we cannot finish the root rotation that we were working on.
+		if !equality.RootCAEqualStable(expectedRootCA, &cluster.RootCA) {
+			return errors.New("target root rotation has changed - unable to complete previous root rotation")
+		}
+
+		var signerCert []byte
+		if len(cluster.RootCA.RootRotation.CAKey) > 0 {
+			signerCert = cluster.RootCA.RootRotation.CACert
+		}
+		// we don't actually have to parse out the default node expiration from the cluster - we are just using
+		// the ca.RootCA object to generate new tokens and the digest
+		updatedRootCA, err := NewRootCA(cluster.RootCA.RootRotation.CACert, signerCert, cluster.RootCA.RootRotation.CAKey,
+			DefaultNodeCertExpiration, nil)
+		if err != nil {
+			return errors.Wrap(err, "invalid cluster root rotation object")
+		}
+		cluster.RootCA = api.RootCA{
+			CACert:     cluster.RootCA.RootRotation.CACert,
+			CAKey:      cluster.RootCA.RootRotation.CAKey,
+			CACertHash: updatedRootCA.Digest.String(),
+			JoinTokens: api.JoinTokens{
+				Worker:  GenerateJoinToken(&updatedRootCA),
+				Manager: GenerateJoinToken(&updatedRootCA),
+			},
+			LastForcedRotation: cluster.RootCA.LastForcedRotation,
+		}
+		return store.UpdateCluster(tx, cluster)
+	})
+}
+
+func (s *Server) reconileNodeRootsAndCerts() {
+	s.mu.Lock()
+	if !s.isRunning() || s.rootCAReconciliationInProgress {
+		s.mu.Unlock()
+		return
+	}
+	ctx := s.ctx
+	s.rootCAReconciliationInProgress = true
+	s.mu.Unlock()
+
+	s.wg.Add(1)
+	logger := log.G(ctx).WithField("method", "(*Server).reconileNodeRootsAndCerts")
+
+	go func(retryInterval time.Duration) {
+		defer func() {
+			s.wg.Done()
+			s.mu.Lock()
+			s.rootCAReconciliationInProgress = false
+			s.mu.Unlock()
+		}()
+
+		for {
```

Comment:
I have created a mapping of all the nodes that is updated by events.  On a root CA update, if the issuer has changed and there's a root rotation present, it will start a reconciliation loop similar to the previous one, except it no longer polls the store, it just checks its in-memory map of nodes.